kurtosis-devnet/tests/eigenda/failover_test.go

```diff
@@ -0,0 +1,409 @@
+	GethL1Endpoint       string `kurtosis:"el-1-geth-teku,rpc"`
+	EigendaProxyEndpoint string `kurtosis:"da-server-op-kurtosis,http"`
+	// Adding new endpoints is as simple as adding a new field with a kurtosis tag
+	// NewServiceEndpoint   string `kurtosis:"new-service-name,port-name"`
+}
+
+func getEndpointsFromKurtosis(enclaveCtx *enclaves.EnclaveContext) (*EnclaveServiceEndpoints, error) {
+	endpoints := &EnclaveServiceEndpoints{}
+
+	// Get the type of the struct to iterate over fields
+	t := reflect.TypeOf(endpoints).Elem()
+	v := reflect.ValueOf(endpoints).Elem()
+
+	// Iterate over all fields in the struct
+	for i := 0; i < t.NumField(); i++ {
+		field := t.Field(i)
+
+		// Get the kurtosis tag
+		tag := field.Tag.Get("kurtosis")
+		if tag == "" {
+			continue // Skip fields without tags
+		}
+
+		// Parse the tag to get service name and port name
+		parts := strings.Split(tag, ",")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid kurtosis tag format for field %s: %s", field.Name, tag)
+		}
+
+		serviceName := parts[0]
+		portName := parts[1]
+
+		// Get the service context
+		serviceCtx, err := enclaveCtx.GetServiceContext(serviceName)
+		if err != nil {
+			return nil, fmt.Errorf("GetServiceContext for %s: %w", serviceName, err)
+		}
+
+		// Get the port
+		port, ok := serviceCtx.GetPublicPorts()[portName]
+		if !ok {
+			return nil, fmt.Errorf("service %s doesn't expose %s port", serviceName, portName)
+		}
+
+		// Set the endpoint URL in the struct field
+		endpoint := fmt.Sprintf("http://localhost:%d", port.GetNumber())
+		v.Field(i).SetString(endpoint)
+	}
+
+	return endpoints, nil
+}
+
+type EnclaveServiceClients struct {
+	opNodeClient         *rpc.Client
+	gethL1Client         *ethclient.Client
+	proxyMemconfigClient *ProxyMemconfigClient
+}
+
+func getClientsFromEndpoints(endpoints *EnclaveServiceEndpoints) (*EnclaveServiceClients, error) {
+	opNodeClient, err := rpc.Dial(endpoints.OpNodeEndpoint)
+	if err != nil {
+		return nil, fmt.Errorf("rpc.Dial: %w", err)
+	}
+
+	gethL1Client, err := ethclient.Dial(endpoints.GethL1Endpoint)
+	if err != nil {
+		return nil, fmt.Errorf("ethclient.Dial: %w", err)
+	}
+
+	proxyMemconfigClient := &ProxyMemconfigClient{
+		Client: memconfig_client.New(&memconfig_client.Config{URL: endpoints.EigendaProxyEndpoint}),
+	}
+
+	return &EnclaveServiceClients{
+		opNodeClient:         opNodeClient,
+		gethL1Client:         gethL1Client,
+		proxyMemconfigClient: proxyMemconfigClient,
+	}, nil
+}
+
+// ProxyMemconfigClient is a wrapper around the memconfig client that adds a Failover method
+// TODO: we should upstream this to eigenda-proxy repo
+type ProxyMemconfigClient struct {
+	*memconfig_client.Client
+}
+
+// Update the proxy's memstore config to start returning 503 errors
+// Note: we have to GetConfig, update it and then UpdateConfig because the client doesn't implement a "patch" method,
+//
+//	even though the API does support it.
```

Comment:
nit: weird formatting here, with newline and tab mid-sentence

Reply:
[85b5c12](https://github.com/Layr-Labs/optimism/pull/33/commits/85b5c12b4bf96e8e8e062467a1e29df343ef5436)